### PR TITLE
Sourcery Starbot ⭐ refactored rodrigosnader/deepdow

### PR DESCRIPTION
**Pull request by SourceryAI**
Thanks for starring [sourcery-ai/sourcery](/sourcery-ai/sourcery) ✨ 🌟 ✨

Here's your pull request refactoring your most popular Python repo.

If you want Sourcery to refactor all your Python repos and incoming pull requests [install our bot](/apps/sourcery-ai/installations/new).

<details>
<summary>Review changes via command line</summary>
<p>To manually merge these changes, make sure you're on the <code>master</code> branch, then run:</p>
<pre>
git fetch https://github.com/sourcery-ai-bot/deepdow master
git merge --ff-only FETCH_HEAD
git reset HEAD^
</pre>
</details>

### CODE REVIEW

File: deepdow/benchmarks.py [hunk 1]
```diff
@@ -61,9 +61,7 @@ def __call__(self, x):
         x_rets = x[:, self.returns_channel, ...]
         vols = x_rets.std(dim=1) if self.use_std else x_rets.var(dim=1)
         ivols = 1 / (vols + eps)
-        weights = ivols / ivols.sum(dim=1, keepdim=True)
-
-        return weights
+        return ivols / ivols.sum(dim=1, keepdim=True)
```
Comment (SourceryAI) [inline deepdow/benchmarks.py:64]: Function `InverseVolatility.__call__` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/callbacks.py [hunk 2]
```diff
@@ -217,7 +217,10 @@ def on_epoch_end(self, metadata):
         epoch = metadata['epoch']
         stats = self.run.history.metrics_per_epoch(epoch)
 
-        if not (len(stats['lookback'].unique()) == 1 and len(stats['model'].unique()) == 1):
+        if (
+            len(stats['lookback'].unique()) != 1
+            or len(stats['model'].unique()) != 1
+        ):
```
Comment (SourceryAI) [inline deepdow/callbacks.py:223]: Function `EarlyStoppingCallback.on_epoch_end` refactored with the following changes:
- Simplify logical expression using De Morgan identities ([`de-morgan`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#de-morgan))

File: deepdow/callbacks.py [hunk 3]
```diff
@@ -398,7 +401,10 @@ def on_epoch_end(self, metadata):
...
```
Comment (SourceryAI) [inline deepdow/callbacks.py:407]: Function `ModelCheckpointCallback.on_epoch_end` refactored with the following changes:
- Simplify logical expression using De Morgan identities ([`de-morgan`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#de-morgan))

File: deepdow/experiments.py [hunk 4]
```diff
@@ -79,11 +79,7 @@ def pretty_print(self, epoch=None):
             If epoch given, then a results only over this epoch. If epoch is `None` then print results over all epochs.
 
         """
-        if epoch is None:
-            df = self.metrics
-
-        else:
-            df = self.metrics_per_epoch(epoch)
+        df = self.metrics if epoch is None else self.metrics_per_epoch(epoch)
```
Comment (SourceryAI) [inline deepdow/experiments.py:82]: Function `History.pretty_print` refactored with the following changes:
- Replace if statement with if expression ([`assign-if-exp`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#assign-if-exp))

File: deepdow/experiments.py [hunk 5]
```diff
@@ -172,7 +168,7 @@ def __init__(self, network, loss, train_dataloader, val_dataloaders=None, metric
             pass
 
         elif isinstance(metrics, dict):
-            if not all([isinstance(x, Loss) for x in metrics.values()]):
+            if not all(isinstance(x, Loss) for x in metrics.values()):
```
Comment (SourceryAI) [inline deepdow/experiments.py:171]: Function `Run.__init__` refactored with the following changes:
- Replace unneeded comprehension with generator ([`comprehension-to-generator`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#comprehension-to-generator))

File: deepdow/losses.py [hunk 6]
```diff
@@ -29,9 +29,7 @@ def covariance(x, y):
     xm = x - mean_x  # (n_samples, horizon)
     ym = y - mean_y  # (n_samples, horizon)
 
-    cov = (xm * ym).sum(dim=1) / horizon
-
-    return cov
+    return (xm * ym).sum(dim=1) / horizon
```
Comment (SourceryAI) [inline deepdow/losses.py:32]: Function `covariance` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/losses.py [hunk 7]
```diff
@@ -353,16 +351,16 @@ def __pow__(self, power):
...
-        else:
+        if not isinstance(power, (int, float)):
             raise TypeError('Unsupported type: {}'.format(type(power)))
 
+        new_instance = Loss()
+        new_instance._call = MethodType(lambda inst, weights, y: self(weights, y) ** power, new_instance)
+        new_instance._repr = MethodType(lambda inst: '({}) ** {}'.format(self.__repr__(), power),
+                                        new_instance)
+
+        return new_instance
```
Comment (SourceryAI) [inline deepdow/losses.py:362]: Function `Loss.__pow__` refactored with the following changes:
- Swap if/else branches ([`swap-if-else-branches`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#swap-if-else-branches))
- Remove unnecessary else after guard condition ([`remove-unnecessary-else`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-unnecessary-else))

File: deepdow/nn.py [hunk 8]
```diff
@@ -153,10 +153,7 @@ def forward(self, x):
         gamma_sqrt_all = torch.ones(len(x)).to(device=x.device, dtype=x.dtype) * self.gamma_sqrt
         alpha_all = torch.ones(len(x)).to(device=x.device, dtype=x.dtype) * self.alpha
 
-        # weights
-        weights = self.portfolio_opt_layer(exp_rets, covmat, gamma_sqrt_all, alpha_all)
-
-        return weights
+        return self.portfolio_opt_layer(exp_rets, covmat, gamma_sqrt_all, alpha_all)
```
Comment (SourceryAI) [inline deepdow/nn.py:156]: Function `BachelierNet.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

This removes the following comments ( [why?](https://github.com/sourcery-ai/sourcery/wiki/FAQ#comment-removal) ):
```
# weights
```

File: deepdow/nn.py [hunk 9]
```diff
@@ -262,9 +259,7 @@ def __call__(self, x):
 
         temperatures = torch.ones(n_samples).to(device=x.device, dtype=x.dtype) * self.temperature
 
-        weights = self.portfolio_opt_layer(x, temperatures)
-
-        return weights
+        return self.portfolio_opt_layer(x, temperatures)
```
Comment (SourceryAI) [inline deepdow/nn.py:262]: Function `KeynesNet.__call__` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/nn.py [hunk 10]
```diff
@@ -352,9 +347,7 @@ def forward(self, x):
         x = self.linear(x)
 
         temperatures = torch.ones(n_samples).to(device=x.device, dtype=x.dtype) * self.temperature
-        weights = self.allocate_layer(x, temperatures)
-
-        return weights
+        return self.allocate_layer(x, temperatures)
```
Comment (SourceryAI) [inline deepdow/nn.py:350]: Function `LinearNet.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/nn.py [hunk 11]
```diff
@@ -474,9 +467,9 @@ def forward(self, x):
         gamma_all = torch.ones(len(x)).to(device=x.device, dtype=x.dtype) * self.gamma_sqrt
         alpha_all = torch.ones(len(x)).to(device=x.device, dtype=x.dtype) * self.alpha
 
-        weights = self.portfolio_opt_layer(exp_returns_all, covariance_all, gamma_all, alpha_all)
-
-        return weights
+        return self.portfolio_opt_layer(
+            exp_returns_all, covariance_all, gamma_all, alpha_all
+        )
```
Comment (SourceryAI) [inline deepdow/nn.py:472]: Function `ThorpNet.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/utils.py [hunk 12]
```diff
@@ -108,7 +108,7 @@ def check_indices_agree(*frames):
             If indices/colums do not agree.
 
         """
-        if not all([isinstance(x, (pd.Series, pd.DataFrame)) for x in frames]):
+        if not all(isinstance(x, (pd.Series, pd.DataFrame)) for x in frames):
```
Comment (SourceryAI) [inline deepdow/utils.py:111]: Function `PandasChecks.check_indices_agree` refactored with the following changes:
- Replace unneeded comprehension with generator ([`comprehension-to-generator`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#comprehension-to-generator))

File: deepdow/visualize.py [hunk 13]
```diff
@@ -74,9 +74,7 @@ def generate_metrics_table(benchmarks, dataloader, metrics, device=None, dtype=N
                                                  'metric': metric_name,
                                                  'value': metric_per_s}))
 
-    metrics_table = pd.concat(all_entries)
-
-    return metrics_table
+    return pd.concat(all_entries)
```
Comment (SourceryAI) [inline deepdow/visualize.py:77]: Function `generate_metrics_table` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/visualize.py [hunk 14]
```diff
@@ -142,9 +140,10 @@ def generate_cumrets(benchmarks, dataloader, device=None, dtype=None, returns_ch
             all_entries[bm_name].append(pd.DataFrame(cumrets.detach().cpu().numpy(),
                                                      index=timestamps))
 
-    cumrets_dict = {bm_name: pd.concat(entries).sort_index() for bm_name, entries in all_entries.items()}
-
-    return cumrets_dict
+    return {
+        bm_name: pd.concat(entries).sort_index()
+        for bm_name, entries in all_entries.items()
+    }
```
Comment (SourceryAI) [inline deepdow/visualize.py:146]: Function `generate_cumrets` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/data/load.py [hunk 15]
```diff
@@ -131,9 +131,6 @@ def collate_uniform(batch, n_assets_range=(5, 10), lookback_range=(2, 20), horiz
     if asset_ixs is None:
         n_assets = torch.randint(low=n_assets_range[0], high=min(n_assets_max + 1, n_assets_range[1]), size=(1,))[0]
         asset_ixs = torch.multinomial(torch.ones(n_assets_max), n_assets.item(), replacement=False)
-    else:
-        pass
-
```
Comment (SourceryAI) [inline deepdow/data/load.py:136]: Function `collate_uniform` refactored with the following changes:
- Remove redundant pass statement ([`remove-redundant-pass`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-redundant-pass))

File: deepdow/layers/allocate.py [hunk 16]
```diff
@@ -158,9 +158,7 @@ def forward(self, covmat, rets=None):
 
             w_l.append(w_final)
 
-        res = torch.stack(w_l, dim=0)
-
-        return res
+        return torch.stack(w_l, dim=0)
```
Comment (SourceryAI) [inline deepdow/layers/allocate.py:161]: Function `NCO.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/layers/allocate.py [hunk 17]
```diff
@@ -475,9 +473,7 @@ def __init__(self, n_assets, temperature=1, max_weight=1):
         x = cp.Parameter(n_assets)
         w = cp.Variable(n_assets)
         obj = cp.sum_squares(x - w)
-        cons = [cp.sum(w) == 1,
-                0. <= w,
-                w <= max_weight]
+        cons = [cp.sum(w) == 1, w >= 0., w <= max_weight]
```
Comment (SourceryAI) [inline deepdow/layers/allocate.py:476]: Function `SparsemaxAllocator.__init__` refactored with the following changes:
- Ensure constant in comparison is on the right ([`flip-comparison`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#flip-comparison))

File: deepdow/layers/misc.py [hunk 18]
```diff
@@ -25,9 +25,7 @@ def forward(self, covmat):
         stds = torch.sqrt(torch.diagonal(covmat, dim1=1, dim2=2))
         stds_ = stds.view(n_samples, n_assets, 1)
 
-        corr = covmat / torch.matmul(stds_, stds_.permute(0, 2, 1))
-
-        return corr
+        return covmat / torch.matmul(stds_, stds_.permute(0, 2, 1))
```
Comment (SourceryAI) [inline deepdow/layers/misc.py:28]: Function `Cov2Corr.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/layers/misc.py [hunk 19]
```diff
@@ -53,9 +51,12 @@ def __init__(self, sqrt=True, shrinkage_strategy='diagonal', shrinkage_coef=0.5)
...
 
-        if shrinkage_strategy is not None:
-            if shrinkage_strategy not in {'diagonal', 'identity', 'scaled_identity'}:
-                raise ValueError('Unrecognized shrinkage strategy {}'.format(shrinkage_strategy))
+        if shrinkage_strategy is not None and shrinkage_strategy not in {
+            'diagonal',
+            'identity',
+            'scaled_identity',
+        }:
+            raise ValueError('Unrecognized shrinkage strategy {}'.format(shrinkage_strategy))
```
Comment (SourceryAI) [inline deepdow/layers/misc.py:59]: Function `CovarianceMatrix.__init__` refactored with the following changes:
- Merge nested if conditions ([`merge-nested-ifs`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#merge-nested-ifs))

File: deepdow/layers/transform.py [hunk 20]
```diff
@@ -171,15 +171,13 @@ def forward(self, x, tform):
...
 
-        x_warped = nn.functional.grid_sample(x,
+        return nn.functional.grid_sample(x,
                                              grid,
                                              mode=self.mode,
                                              padding_mode=self.padding_mode,
                                              align_corners=True,
                                              )
 
-        return x_warped
```
Comment (SourceryAI) [inline deepdow/layers/transform.py:181]: Function `Warp.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: deepdow/layers/transform.py [hunk 21]
```diff
@@ -234,11 +232,9 @@ def forward(self, x, scale):
...
         grid = nn.functional.affine_grid(theta, x.shape, align_corners=True)
-        x_zoomed = nn.functional.grid_sample(x,
+        return nn.functional.grid_sample(x,
                                              grid,
                                              mode=self.mode,
                                              padding_mode=self.padding_mode,
                                              align_corners=True,
                                              )
-
-        return x_zoomed
```
Comment (SourceryAI) [inline deepdow/layers/transform.py:244]: Function `Zoom.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: examples/end_to_end/getting_started.py [hunk 22]
```diff
@@ -150,9 +150,7 @@ def forward(self, x):
         x = self.dense_layer(x)
 
         temperatures = torch.ones(n_samples).to(device=x.device, dtype=x.dtype) * self.temperature
-        weights = self.allocate_layer(x, temperatures)
-
-        return weights
+        return self.allocate_layer(x, temperatures)
```
Comment (SourceryAI) [inline examples/end_to_end/getting_started.py:153]: Function `GreatNet.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: examples/end_to_end/iid.py [hunk 23]
```diff
@@ -363,9 +363,7 @@ def forward(self, x):
                                            cov_sqrt[None, ...],
                                            self.gamma_sqrt,
                                            torch.zeros(1).to(device=x.device, dtype=x.dtype))
-        weights_filled = torch.repeat_interleave(weights, n, dim=0)
-
-        return weights_filled
+        return torch.repeat_interleave(weights, n, dim=0)
```
Comment (SourceryAI) [inline examples/end_to_end/iid.py:366]: Function `Net.forward` refactored with the following changes:
- Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#inline-immediately-returned-variable))

File: tests/test_layers.py [hunk 24]
```diff
@@ -481,12 +481,26 @@ def test_n_parameters(self, n_channels, hidden_size, cell_type, bidirectional):
...
-                    (n_channels * hidden_size_a) + (hidden_size_a * hidden_size_a) + 2 * hidden_size_a)
+            assert n_parameters == (
+                n_dir
+                * 4
+                * (
+                    n_channels * hidden_size_a
+                    + hidden_size_a ** 2
+                    + 2 * hidden_size_a
+                )
+            )
```
Comment (SourceryAI) [inline tests/test_layers.py:503]: Function `TestRNN.test_n_parameters` refactored with the following changes:
- Replace x * x with x ** 2 ([`square-identity`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#square-identity))
